File: haskell/basics/quicksort.hs
-- Sort a fixed list with the classic quicksort written as two list comprehensions.
-- Run: runghc quicksort.hs

quicksort :: (Ord a) => [a] -> [a]
quicksort [] = []
quicksort (pivot : rest) =
  quicksort [x | x <- rest, x <= pivot] ++ [pivot] ++ quicksort [x | x <- rest, x > pivot]

main :: IO ()
main = putStrLn . unwords . map show $ quicksort [5, 3, 8, 4, 2, 7, 1, 10, 9, 6 :: Int]

File: elixir/basics/quicksort.exs
# Sort a fixed list with a quicksort written as two clauses over list patterns.
# Run: elixir quicksort.exs

defmodule Quicksort do
  def sort([]), do: []

  def sort([pivot | rest]) do
    smaller = for x <- rest, x <= pivot, do: x
    larger = for x <- rest, x > pivot, do: x
    sort(smaller) ++ [pivot] ++ sort(larger)
  end
end

[5, 3, 8, 4, 2, 7, 1, 10, 9, 6]
|> Quicksort.sort()
|> Enum.join(" ")
|> IO.puts()
